cylc/flow/cfgspec/suite.py

```diff
@@ -1360,126 +1357,75 @@ def upg(cfg, descr):
+    may cause this in cases that we cannot catch here, we might as well
+    return errors that we can catch here.
+    """
+    if 'runtime' not in cfg:
+        # this whole function is pointless
+        return
+
+    forbidden_with_platform = [
+        ('remote', 'host'),
+        ('job', 'batch system'),
+        ('job', 'batch submit command template')
+    ]
+    warnlines = ''
+    limit_count = 0
+    for task_name, task_cfg in cfg['runtime'].items():
+        for section, key in forbidden_with_platform:
             if (
-                'host' in task_spec_remote and
-                REC_COMMAND.match(task_spec['remote']['host'])
+                section in task_cfg and
+                key in task_cfg[section] and
+                task_cfg[section][key]
             ):
-                LOG.debug(
-                    f"The host setting of '{task_name}' is a function: "
-                    f"Cylc will try to upgrade this task on job submission."
-                )
-                continue
-
-            # Attempt to use the reverse lookup
-            try:
-                platform = platform_from_job_info(
-                    glbl_cfg(cached=False).get(['platforms']),
-                    task_spec_job,
-                    task_spec_remote
-                )
-            except PlatformLookupError as exc:
-                raise PlatformLookupError(f"for task {task_name}: {exc}")
-            else:
-                # Set platform in config
-                cfg['runtime'][task_name].update({'platform': platform})
-                LOG.warning(f"Platform {platform} auto selected from ")
-                # Remove deprecated items from config
-                for old_spec_item in forbidden_with_platform:
-                    for task_section in ['job', 'remote']:
-                        if (
-                            task_section in cfg['runtime'][task_name] and
-                            old_spec_item in
-                                cfg['runtime'][task_name][task_section].keys()
-                        ):
-                            poppable = cfg['runtime'][task_name][task_section]
-                            poppable.pop(old_spec_item)
-                    LOG.warning(
-                        f"Cylc 7 {old_spec_item} removed."
+                if 'platform' in task_cfg:
+                    # We can be sure that this is wrong, although due to
+                    # inheritance we cannot catch all cases here.
+                    LOG.error(
+                        "A mixture of Cylc 7 (host) and Cylc 8 (platform)"
+                        " logic should not be used. In this case the "
+                        "following are not compatible\n"
+                        f"[{task_name}][{section}]{key} = "
+                        f"{task_cfg[section][key]} and "
+                        f"[{task_name}][platform] = {task_cfg['platform']}"
+                    )
+                    raise UpgradeError()
```

Comment:
Is there a reason for logging with level "error" and then raising an empty error, instead of just raising the error with the message above?

Reply:
Yes - you get more error and less traceback.

Reply:
Won't you get traceback for the empty exception anyway?

Reply:
Oh - I've not got this right.

Reply:
This, it turned out was a proxy for a fairly nasty problem which @oliver-sanders found a fix for, for me. A special handler for this exception had to be added at about line 1800 in scheduler.py